app/src/org/commcare/UpdateToPrompt.java

```diff
@@ -0,0 +1,154 @@
+        if (forceByDate != null) {
+            this.forceByDate = DateUtils.parseDate(forceByDate);
+        }
+        this.isApkUpdate = isApkUpdate;
+        this.versionString = version;
+        buildFromVersionString();
+    }
+
+    public UpdateToPrompt() {
+        // for deserialization
+    }
+
+    private void buildFromVersionString() {
+        if (isApkUpdate) {
+            this.apkVersion = new ApkVersion(versionString);
+        } else {
+            this.cczVersion = Integer.parseInt(versionString);
+        }
+    }
+
+    public boolean isPastForceByDate() {
+        return forceByDate != null && (forceByDate.getTime() < System.currentTimeMillis());
+    }
+
+    public void registerWithSystem() {
+        CommCareApp currentApp = CommCareApplication.instance().getCurrentApp();
+        if (isNewerThanCurrentVersion(currentApp)) {
+            printDebugStatement();
+            writeToPrefsObject(currentApp.getAppPreferences());
+        } else {
+            // If the latest signal we're getting is that our current version is up-to-date,
+            // then we should wipe any update prompt for this type that was previously stored
+            CommCareHeartbeatManager.wipeStoredUpdate(this.isApkUpdate);
+        }
+    }
+
+    private void printDebugStatement() {
+        if (isApkUpdate) {
+            System.out.println(".apk version to prompt for update set to " + apkVersion);
+        } else {
+            System.out.println(".ccz version to prompt for update set to " + cczVersion);
+        }
+        if (this.forceByDate != null) {
+            System.out.println("force-by date is " + forceByDate);
+        }
+    }
+
+    public boolean isNewerThanCurrentVersion(CommCareApp currentApp) {
+        if (isApkUpdate) {
+            try {
+                Context c = CommCareApplication.instance();
+                PackageInfo pi = c.getPackageManager().getPackageInfo(c.getPackageName(), 0);
+                ApkVersion currentVersion = new ApkVersion(pi.versionName);
+                return currentVersion.compareTo(this.apkVersion) < 0;
+            } catch (PackageManager.NameNotFoundException e) {
+                // This shouldn't happen, but it if it does, there's no way for us to know if the
+                // update version is newer, so don't prompt
+                Logger.log(AndroidLogger.TYPE_ERROR_WORKFLOW,
+                        "Couldn't get current .apk version to compare with in UpdateToPrompt: "
+                                + e.getMessage());
+                return false;
+            }
+        } else {
+            int currentVersion = currentApp.getCommCarePlatform().getCurrentProfile().getVersion();
+            return currentVersion < this.cczVersion;
+        }
+    }
+
+    private void writeToPrefsObject(SharedPreferences prefs) {
+        ByteArrayOutputStream baos = null;
+        DataOutputStream serializedStream = null;
```

Comment:
I think there's an existing Utility method for "Pass in an externalizable, get back an byte[]", I'd try to re-use that as the error handling here is quite unfortunate to end up replicating